src/Core/System/Snippet/DataTransfer/Metadata/MetadataCollection.php

```diff
@@ -0,0 +1,88 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\System\Snippet\DataTransfer\Metadata;
+
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Core\Framework\Struct\Collection;
+
+/**
+ * @internal
+ *
+ * @extends Collection<MetadataEntry>
+ */
+#[Package('discovery')]
+class MetadataCollection extends Collection
+{
+    /**
+     * @param list<MetadataEntry> $elements
+     */
+    public function __construct(
+        array $elements = [],
+    ) {
+        $elements = $this->getIndexedList($elements);
+        parent::__construct($elements);
+    }
```

Comment:
if I understand this correctly, you want to have the collection always be indexed by the locale of its metadataEntry. 

If so, I would suggest something different, because your are not considering the `set` and `add` methods. 

```suggestion
    public function set($key, $element): void
    {
        parent::set($element->locale, $element);
    }

    public function add($element): void
    {
        $this->validateType($element);

        parent::set($element->locale, $element);
    }
```

the construct case is also considered, as it is calling the `set` method

Reply:
i like the idea 👍